### PR TITLE
Add support for reading json in multipart/form-data requests

### PR DESCRIPTION
**Pull request by pranavkm**
POC for what this would look like. @davidfowl wasn't very happy about adding API to `IFormCollection`, but wanted to add something more full featured that `IFormCollection` could read from instead. Perhaps `IMultipartFormCollection`. Mvc's implementation for the multipart binder makes me a bit sad, at least it wasn't too difficult to get going.

### CODE REVIEW

File: src/Http/Http/src/Features/FormFeature.cs [hunk 1]
```diff
@@ -177,6 +178,8 @@ private async Task<IFormCollection> InnerReadFormAsync(CancellationToken cancell
                             throw new InvalidDataException("Form section has invalid Content-Disposition value: " + section.ContentDisposition);
                         }
 
+                        contentTypesAccumulator.Append(contentDisposition.Name.Value, section.ContentType);
```
Comment (Tratcher) [inline src/Http/Http/src/Features/FormFeature.cs:181]: Move this under the IsFormDisposition case. Files already give you access to their headers, and you don't want to track this for 'other' types that we discarded.

File: src/Http/Http.Features/src/IFormCollection.cs [hunk 2]
```diff
@@ -90,5 +90,13 @@ public interface IFormCollection : IEnumerable<KeyValuePair<string, StringValues
         /// </summary>
         /// <returns>The files included with the request.</returns>
         IFormFileCollection Files { get; }
+
+#if NETCOREAPP
+        bool TryGetContentType(string sectionName, out StringValues contentType)
```
Comment (Tratcher) [inline src/Http/Http.Features/src/IFormCollection.cs:95]: content-type should just be string, it's a single value header. We do the same thing on HttpRequest/Response.

File: src/Mvc/Mvc.Core/src/ModelBinding/Binders/MultipartSectionModelBinder.cs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+                var bytes = Encoding.UTF8.GetBytes(value);
+                request.Body = new MemoryStream(bytes);
+                request.ContentType = multipartContentType.MediaType.Value;
+
+                await _worker.ExecuteAsync(bindingContext, modelBindingKey);
+            }
+            finally
+            {
+                request.Body = originalBody;
+                request.ContentType = requestContentType;
```
Comment (Tratcher) [inline src/Mvc/Mvc.Core/src/ModelBinding/Binders/MultipartSectionModelBinder.cs:104]: Ok, this part is wacky, you shouldn't be messing with the request like that. If you really need an HttpRequest then create a new DefaultHttpContext.

Comment (davidfowl) [reply src/Mvc/Mvc.Core/src/ModelBinding/Binders/MultipartSectionModelBinder.cs:104]: Why change the body at all?

File: src/Http/Http.Features/src/IFormCollection.cs [hunk 4]
```diff
@@ -90,5 +90,13 @@ public interface IFormCollection : IEnumerable<KeyValuePair<string, StringValues
         /// </summary>
         /// <returns>The files included with the request.</returns>
         IFormFileCollection Files { get; }
+
+#if NETCOREAPP
```
Comment (davidfowl) [inline src/Http/Http.Features/src/IFormCollection.cs:94]: I don't think we should do this.

File: src/Mvc/Mvc.Core/src/ModelBinding/Binders/MultipartSectionModelBinder.cs [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+            }
+
+            _logger?.AttemptingToBindModel(bindingContext);
+
+            var originalBody = request.Body;
+            var value = values[0];
+
+            try
+            {
+                var bytes = Encoding.UTF8.GetBytes(value);
```
Comment (davidfowl) [inline src/Mvc/Mvc.Core/src/ModelBinding/Binders/MultipartSectionModelBinder.cs:95]: 😢

File: src/Mvc/Mvc.Core/src/ModelBinding/Binders/MultipartSectionModelBinder.cs (hunk 3 above)
Comment (pranavkm) [reply src/Mvc/Mvc.Core/src/ModelBinding/Binders/MultipartSectionModelBinder.cs:104]: > Why change the body at all?

Input formatters read the body. Unless we want to start adding special behavior in every one of those. 

> If you really need an HttpRequest then create a new DefaultHttpContext.

I don't mind doing this, but is there a reason that's preferred over changing the current request? Code hygiene?

Comment (Tratcher) [reply src/Mvc/Mvc.Core/src/ModelBinding/Binders/MultipartSectionModelBinder.cs:104]: Assigning the body has some odd side-effects like adding a feature wrapper.

Input formatters really should be able to read an arbitrary stream.

**Comment by davidfowl** [comment]
I don’t feel good about this solution, as simple as it is for a couple of reasons:
- The new default interface method on net5 only. This assembly is really only cross compiled because of IFeatureCollection and friends not necessarily individual features.
- Large JSON payloads may end up on the LOH. Now that’s no different from an individual form value but it’s more likely with a complex JSON object graph.

The current interface assumes the entire form is read and buffered up front (even the files are buffered in memory or on disk). If this becomes the way we recommend doing complex multi-part requests then I’d like to see some performance tests to show that we understand the issues.

**Comment by Tratcher** [comment]
Design proposal:
```
    public interface IMultipartForm : IFormCollection
    {
        bool TryGetSection(string key, out MultipartSection value);
    }
```

Re-implement the IFormCollection indexer and TryGetValue to lazily decode from TryGetSection. Count, Keys, ContainsKey, and Files can all remain pre-evaluated properties as none of them have to evaluate the body.

FormFeature can enable buffering even for non-file sections and add the raw sections to the detailed form collection.

MVC does its normal thing to read and process the form, but downcasts to IMultipartForm if it's trying to model bind for these new scenarios.

On second thought @davidfowl, does IMultipartForm need to derive from IFormCollection? Or would you just have the same type implement both in this example? The Count and Keys properties overlap, but there's no reason a lower level IMultipartForm would necessarily need the other APIs.

**Comment by davidfowl** [comment]
@pranavkm did we file an issue to explore this? Did we merge something else?

**Comment by pranavkm** [comment]
We haven't merged it as yet. We still have an issue for this https://github.com/dotnet/aspnetcore/issues/4868 that hopefully we get a better stab at in 6.0.

**Pull request closed without merging**